Running after a previous failure can cause a error related to cached residue definitions

I ran the 7jjf test case once and it failed due to having an incompatible version of openff-units. Then I installed the new version of openff-units, and the test case failed with something related to caching the CCD residue definitions:

```
ImportError while loading conftest '/Users/jeffreywagner/projects/OpenForceField/openff-pablo/openff/pablo/_tests/conftest.py'.
../__init__.py:7: in <module>
    from . import ccd, chem, exceptions, residue
../ccd/__init__.py:11: in <module>
    from . import patches
../ccd/patches.py:9: in <module>
    from openff.pablo.chem import DISULFIDE_BOND
../__init__.py:7: in <module>
    from . import ccd, chem, exceptions, residue
../ccd/__init__.py:28: in <module>
    CCD_RESIDUE_DEFINITION_CACHE: CcdCache = CcdCache(
../ccd/_ccdcache.py:68: in __init__
    self._add_definition_from_str(file.read_text())
../ccd/_ccdcache.py:119: in _add_definition_from_str
    definition = self._res_def_from_ccd_str(s)
../ccd/_ccdcache.py:229: in _res_def_from_ccd_str
    return inner(s)
../ccd/_ccdcache.py:167: in inner
    linking_bond = LINKING_TYPES[linking_type]
E   KeyError: 'DNA LINKING'
```

--- a/devtools/conda-envs/test_env.yaml
+++ b/devtools/conda-envs/test_env.yaml
@@ -37,5 +37,7 @@ dependencies:
   # Data
   - numpy
 
-  #- pip:
-  #  - codecov
+  - pip:
+      # - codecov
+      - git+https://github.com/openforcefield/openff-toolkit.git@better_typing
+      - git+https://github.com/openforcefield/openff-units.git@better_typing

--- a/openff/pablo/_pdb.py
+++ b/openff/pablo/_pdb.py
@@ -493,7 +493,7 @@ def _add_to_molecule(
                 for old_idx, atom in enumerate(this_molecule.atoms):
                     old_to_new[old_idx] = other_molecule._add_atom(
                         atomic_number=atom.atomic_number,
-                        formal_charge=atom.formal_charge.m,
+                        formal_charge=atom.formal_charge.m,  # type:ignore
                         is_aromatic=atom.is_aromatic,
                         stereochemistry=atom.stereochemistry,
                         name=atom.name,

--- a/openff/pablo/_pdb.py
+++ b/openff/pablo/_pdb.py
@@ -407,10 +407,6 @@ def _add_to_molecule(
     for pdb_index in sorted(residue_match.res_atom_idcs):
         atom_def = residue_match.atom(pdb_index)
 
-        if data.alt_loc[pdb_index] != "":
-            # TODO: Support altlocs (probably in PdbData, maybe PdbData.residues()?)
-            raise ValueError("altloc not yet supported")
-
         mol_atom_idx = this_molecule._add_atom(
             atomic_number=elements.NUMBERS[atom_def.symbol],
             formal_charge=atom_def.charge,

--- a/openff/pablo/_pdb_data.py
+++ b/openff/pablo/_pdb_data.py
@@ -8,7 +8,10 @@ from pathlib import Path
 from typing import Any, DefaultDict, Self
 
 from ._utils import __UNSET__, dec_hex, int_or_none, with_neighbours
-from .exceptions import UnknownOrAmbiguousSerialInConectError
+from .exceptions import (
+    NoMatchingResidueDefinitionError,
+    UnknownOrAmbiguousSerialInConectError,
+)
 from .residue import AtomDefinition, ResidueDefinition
 
 __all__ = [
@@ -462,7 +465,21 @@ class PdbData:
                     if match is not None:
                         residue_matches.append(match)
 
+            if len(residue_matches) == 0:
+                raise NoMatchingResidueDefinitionError(
+                    res_atom_idcs,
+                    self,
+                    unknown_molecules=[],
+                    additional_substructures=additional_substructures,
+                    residue_database=residue_database,
+                    verbose_errors=True,
+                )
+
             name_matches.append(residue_matches)
+
+        print(
+            f"{[[match.residue_definition.description for match in matches] for matches in name_matches]=}",
+        )
 
         # Check for polymer bonds
         prev_filtered_matches: list[ResidueMatch] = []
@@ -512,6 +529,10 @@ class PdbData:
 
             linkage_matches.append(this_filtered_matches)
             prev_filtered_matches = this_filtered_matches
+
+        print(
+            f"{[[match.residue_definition.description for match in matches] for matches in linkage_matches]=}",
+        )
 
         # Check for crosslinks
         # TODO: This could be simplified if we required crosslinking atoms not to have synonyms

--- a/openff/pablo/_pdb_data.py
+++ b/openff/pablo/_pdb_data.py
@@ -10,7 +10,6 @@ from typing import Any, DefaultDict, Self
 
 from ._utils import __UNSET__, dec_hex, int_or_none, with_neighbours
 from .exceptions import (
-    NoMatchingResidueDefinitionError,
     UnknownOrAmbiguousSerialInConectError,
 )
 from .residue import AtomDefinition, ResidueDefinition
@@ -485,16 +484,6 @@ class PdbData:
                     )
                     if match is not None:
                         residue_matches.append(match)
-
-            if len(residue_matches) == 0:
-                raise NoMatchingResidueDefinitionError(
-                    res_atom_idcs,
-                    self,
-                    unknown_molecules=[],
-                    additional_substructures=additional_substructures,
-                    residue_database=residue_database,
-                    verbose_errors=True,
-                )
 
             name_matches.append(residue_matches)
 

--- a/openff/pablo/_pdb_data.py
+++ b/openff/pablo/_pdb_data.py
@@ -1,4 +1,5 @@
 import dataclasses
+import logging
 from collections import defaultdict
 from collections.abc import Iterable, Iterator, Mapping, Sequence
 from dataclasses import dataclass, field
@@ -317,8 +318,9 @@ class PdbData:
     def residue_indices(self) -> Iterator[tuple[int, ...]]:
         indices = []
         prev = None
-        for atom_idx, residue_info in enumerate(
+        for atom_idx, (alt_loc, *residue_info) in enumerate(
             zip(
+                self.alt_loc,
                 self.model,
                 self.res_name,
                 self.chain_id,
@@ -326,6 +328,9 @@ class PdbData:
                 self.i_code,
             ),
         ):
+            if alt_loc != "":
+                # TODO: Support alt locs
+                raise ValueError("Alt loc not supported")
             if prev == residue_info or prev is None:
                 indices.append(atom_idx)
             else:
@@ -345,8 +350,13 @@ class PdbData:
         if len(res_atom_idcs) == 0:
             raise ValueError("cannot match empty res_atom_idcs")
 
+        logging.debug(f"  Attempting match against {residue_definition.description}")
+
         # Skip definitions with too few atoms
         if len(residue_definition.atoms) < len(res_atom_idcs):
+            logging.debug(
+                f"    Match failed: Too few atoms in residue definition ({len(residue_definition.atoms)} < {len(res_atom_idcs)})",
+            )
             return None
 
         # Skip non-(cross)linking definitions with the wrong number of atoms
@@ -358,6 +368,7 @@ class PdbData:
             )
             != len(res_atom_idcs)
         ):
+            logging.debug("    Match failed: No links and wrong number of atoms")
             return None
 
         # Get the map from the canonical names to the indices
@@ -366,12 +377,16 @@ class PdbData:
                 i: residue_definition.name_to_atom[self.name[i]] for i in res_atom_idcs
             }
         except KeyError:
+            logging.debug("    Match failed: Name missing from residue definition")
             return None
 
         matched_atoms = {atom.name for atom in index_to_atomdef.values()}
 
         # Fail to match if any atoms in PDB file got matched to more than one name
         if len(matched_atoms) != len(res_atom_idcs):
+            logging.debug(
+                "    Match failed: Atom definition matched multiple PDB coordinate records",
+            )
             return None
 
         # This assert should be guaranteed by the above
@@ -382,6 +397,7 @@ class PdbData:
             self.element[i] != "" and self.element[i].lower() != atom.symbol.lower()
             for i, atom in index_to_atomdef.items()
         ):
+            logging.debug("    Match failed: Element mismatch")
             return None
 
         # Check that charges match, but tolerate missing columns
@@ -389,6 +405,7 @@ class PdbData:
             self.charge[i] is not None and self.charge[i] != atom.charge
             for i, atom in index_to_atomdef.items()
         ):
+            logging.debug("    Match failed: Charge mismatch")
             return None
 
         missing_atoms = [
@@ -399,6 +416,7 @@ class PdbData:
         # is either entirely present or entirely absent
         missing_atom_names = {atom.name for atom in missing_atoms}
         if any(not atom.leaving for atom in missing_atoms):
+            logging.debug("    Match failed: Missing atom is not leaving atom")
             return None
         elif (
             (
@@ -426,11 +444,13 @@ class PdbData:
                 )
             )
         ):
+            logging.debug("    Match succeeded!")
             return ResidueMatch(
                 index_to_atomdef=index_to_atomdef,
                 residue_definition=residue_definition,
             )
         else:
+            logging.debug("    Match failed: Missing atoms do not specify link")
             return None
 
     def get_residue_matches(
@@ -442,6 +462,7 @@ class PdbData:
         name_matches: list[list[ResidueMatch]] = []
         atom_idx_to_res_idx: dict[int, int] = {}
         for res_idx, res_atom_idcs in enumerate(self.residue_indices):
+            logging.debug(f"Beginning name-based match of {res_atom_idcs}")
             prototype_index = res_atom_idcs[0]
             res_name = self.res_name[prototype_index]
             atom_idx_to_res_idx.update({i: res_idx for i in res_atom_idcs})
@@ -476,10 +497,6 @@ class PdbData:
                 )
 
             name_matches.append(residue_matches)
-
-        print(
-            f"{[[match.residue_definition.description for match in matches] for matches in name_matches]=}",
-        )
 
         # Check for polymer bonds
         prev_filtered_matches: list[ResidueMatch] = []

--- a/openff/pablo/_tests/test_ccd.py
+++ b/openff/pablo/_tests/test_ccd.py
@@ -1,0 +1,42 @@
+import pytest
+
+from openff.pablo.ccd import CCD_RESIDUE_DEFINITION_CACHE
+
+
+@pytest.mark.parametrize(
+    "resname",
+    [
+        "ALA",
+        "ARG",
+        "ASN",
+        "ASP",
+        "CYS",
+        "GLN",
+        "GLU",
+        "GLY",
+        "HIS",
+        "ILE",
+        "LEU",
+        "LYS",
+        "MET",
+        "PHE",
+        "PRO",
+        "SER",
+        "THR",
+        "TRP",
+        "TYR",
+        "VAL",
+        "DG",
+        "DA",
+        "DT",
+        "DC",
+        "G",
+        "A",
+        "U",
+        "C",
+        "ACE",
+        "NME",
+    ],
+)
+def test_ccdcache_can_load_common_residues(resname: str):
+    CCD_RESIDUE_DEFINITION_CACHE[resname]

--- a/openff/pablo/_tests/test_patches.py
+++ b/openff/pablo/_tests/test_patches.py
@@ -1,0 +1,51 @@
+from openff.pablo.ccd.patches import disambiguate_alt_ids
+from openff.pablo.residue import (
+    AtomDefinition,
+    BondDefinition,
+    ResidueDefinition,
+    _skip_residue_definition_validation,
+)
+
+
+def test_disambiguate_alt_ids():
+    with _skip_residue_definition_validation():
+        with_synonyms = ResidueDefinition(
+            atoms=(
+                AtomDefinition.with_defaults(name="H1", symbol="H", synonyms=["H2"]),
+                AtomDefinition.with_defaults(name="H2", symbol="H", synonyms=["H3"]),
+                AtomDefinition.with_defaults(name="O", symbol="O", synonyms=["O1"]),
+            ),
+            bonds=(
+                BondDefinition.with_defaults("O", "H1"),
+                BondDefinition.with_defaults("O", "H2"),
+            ),
+            crosslink=BondDefinition.with_defaults("O", "H1", order=0),
+            description="water with clashing synonyms",
+            linking_bond=BondDefinition.with_defaults("O", "H2", order=0),
+            residue_name="HOH",
+        )
+        res1, res2 = disambiguate_alt_ids(with_synonyms)
+
+    assert res1 == with_synonyms.replace(
+        atoms=(
+            AtomDefinition.with_defaults(name="H1", symbol="H"),
+            AtomDefinition.with_defaults(name="H2", symbol="H"),
+            AtomDefinition.with_defaults(name="O", symbol="O"),
+        ),
+    )
+
+    assert res2.description == with_synonyms.description + " altids"
+    assert set(res2.atoms) == {
+        AtomDefinition.with_defaults(name="H2", symbol="H"),
+        AtomDefinition.with_defaults(name="H3", symbol="H"),
+        AtomDefinition.with_defaults(name="O1", symbol="O"),
+    }
+    assert set(res2.bonds) == {
+        BondDefinition.with_defaults("O1", "H2"),
+        BondDefinition.with_defaults("O1", "H3"),
+    }
+    assert res2.linking_bond == BondDefinition.with_defaults("O1", "H3", order=0)
+    assert res2.crosslink == BondDefinition.with_defaults("O1", "H2", order=0)
+
+    res1._validate()
+    res2._validate()

--- a/openff/pablo/_tests/test_pdb_data.py
+++ b/openff/pablo/_tests/test_pdb_data.py
@@ -278,6 +278,7 @@ class TestPdbData:
             chain_id=["A"] * 6 + ["B"] * 6,
             res_seq=[1] * 8 + [2] * 4,
             i_code=[" "] * 10 + ["A"] * 2,
+            alt_loc=[""] * 12,
         )
 
         assert list(data.residue_indices) == [

--- a/openff/pablo/_tests/test_pdb_files.py
+++ b/openff/pablo/_tests/test_pdb_files.py
@@ -120,6 +120,6 @@ def connectivity_and_atom_order_and_net_residue_charge_and_metadata_matches_lega
     for pablo_res, legacy_res in zip(pablo_top.residues, legacy_top.residues):
         pablo_res_charge, legacy_res_charge = 0, 0
         for pablo_atom, legacy_atom in zip(pablo_res.atoms, legacy_res.atoms):
-            pablo_res_charge += pablo_atom.formal_charge
-            legacy_res_charge += legacy_atom.formal_charge
+            pablo_res_charge += pablo_atom.formal_charge  # type:ignore
+            legacy_res_charge += legacy_atom.formal_charge  # type:ignore
         assert pablo_res_charge == legacy_res_charge

--- a/openff/pablo/_tests/test_residue.py
+++ b/openff/pablo/_tests/test_residue.py
@@ -2,7 +2,12 @@ import pytest
 from openff.toolkit import Molecule
 
 from openff.pablo.chem import DISULFIDE_BOND, PEPTIDE_BOND
-from openff.pablo.residue import AtomDefinition, BondDefinition, ResidueDefinition
+from openff.pablo.residue import (
+    AtomDefinition,
+    BondDefinition,
+    ResidueDefinition,
+    _skip_residue_definition_validation,
+)
 
 
 class TestBondDefinition:
@@ -11,6 +16,28 @@ class TestBondDefinition:
             BondDefinition.with_defaults("H1", "H2")
             == BondDefinition.with_defaults("H2", "H1").flipped()
         )
+
+    def test_replace_atom1(self):
+        with _skip_residue_definition_validation():
+            replaced = PEPTIDE_BOND.replace(atom1="CO")
+
+        assert PEPTIDE_BOND.atom1 == "C"
+        assert replaced.atom1 == "CO"
+        assert PEPTIDE_BOND.atom2 == replaced.atom2
+        assert PEPTIDE_BOND.order == replaced.order
+        assert PEPTIDE_BOND.aromatic == replaced.aromatic
+        assert PEPTIDE_BOND.stereo == replaced.stereo
+
+    def test_replace_atom2(self):
+        with _skip_residue_definition_validation():
+            replaced = PEPTIDE_BOND.replace(atom2="NH")
+
+        assert PEPTIDE_BOND.atom2 == "N"
+        assert replaced.atom2 == "NH"
+        assert PEPTIDE_BOND.atom1 == replaced.atom1
+        assert PEPTIDE_BOND.order == replaced.order
+        assert PEPTIDE_BOND.aromatic == replaced.aromatic
+        assert PEPTIDE_BOND.stereo == replaced.stereo
 
 
 class TestResidueDefinition:
@@ -36,6 +63,78 @@ class TestResidueDefinition:
             13: "HG",
             14: "HXT",
         }
+
+    def test_replace_residue_name(self, cys_def: ResidueDefinition):
+        with _skip_residue_definition_validation():
+            replaced = cys_def.replace(residue_name="CYX")
+
+        assert cys_def.residue_name == "CYS"
+        assert replaced.residue_name == "CYX"
+        assert cys_def.atoms == replaced.atoms
+        assert cys_def.bonds == replaced.bonds
+        assert cys_def.description == replaced.description
+        assert cys_def.crosslink == replaced.crosslink
+        assert cys_def.linking_bond == replaced.linking_bond
+
+    def test_replace_atoms(self, cys_def: ResidueDefinition):
+        with _skip_residue_definition_validation():
+            replaced = cys_def.replace(atoms=())
+
+        assert cys_def.atoms != ()
+        assert replaced.atoms == ()
+        assert cys_def.bonds == replaced.bonds
+        assert cys_def.description == replaced.description
+        assert cys_def.crosslink == replaced.crosslink
+        assert cys_def.linking_bond == replaced.linking_bond
+        assert cys_def.residue_name == replaced.residue_name
+
+    def test_replace_bonds(self, cys_def: ResidueDefinition):
+        with _skip_residue_definition_validation():
+            replaced = cys_def.replace(bonds=())
+
+        assert cys_def.bonds != ()
+        assert replaced.bonds == ()
+        assert cys_def.atoms == replaced.atoms
+        assert cys_def.description == replaced.description
+        assert cys_def.crosslink == replaced.crosslink
+        assert cys_def.linking_bond == replaced.linking_bond
+        assert cys_def.residue_name == replaced.residue_name
+
+    def test_replace_description(self, cys_def: ResidueDefinition):
+        with _skip_residue_definition_validation():
+            replaced = cys_def.replace(description="Hello there")
+
+        assert replaced.description == "Hello there"
+        assert cys_def.description == "CYSTEINE"
+        assert cys_def.atoms == replaced.atoms
+        assert cys_def.bonds == replaced.bonds
+        assert cys_def.crosslink == replaced.crosslink
+        assert cys_def.linking_bond == replaced.linking_bond
+        assert cys_def.residue_name == replaced.residue_name
+
+    def test_replace_crosslink(self, cys_def: ResidueDefinition):
+        with _skip_residue_definition_validation():
+            replaced = cys_def.replace(crosslink=PEPTIDE_BOND)
+
+        assert replaced.crosslink == PEPTIDE_BOND
+        assert cys_def.crosslink == DISULFIDE_BOND
+        assert cys_def.atoms == replaced.atoms
+        assert cys_def.bonds == replaced.bonds
+        assert cys_def.description == replaced.description
+        assert cys_def.linking_bond == replaced.linking_bond
+        assert cys_def.residue_name == replaced.residue_name
+
+    def test_replace_linking_bond(self, cys_def: ResidueDefinition):
+        with _skip_residue_definition_validation():
+            replaced = cys_def.replace(linking_bond=DISULFIDE_BOND)
+
+        assert replaced.linking_bond == DISULFIDE_BOND
+        assert cys_def.linking_bond == PEPTIDE_BOND
+        assert cys_def.atoms == replaced.atoms
+        assert cys_def.bonds == replaced.bonds
+        assert cys_def.description == replaced.description
+        assert cys_def.crosslink == replaced.crosslink
+        assert cys_def.residue_name == replaced.residue_name
 
     def test_from_molecule(
         self,

--- a/openff/pablo/_tests/test_utils.py
+++ b/openff/pablo/_tests/test_utils.py
@@ -104,13 +104,13 @@ def test_assign_stereochemistry_from_3d_bonds():
     cis_mol.generate_conformers(n_conformers=1)
     trans_mol.generate_conformers(n_conformers=1)
 
-    undef_mol.add_conformer(cis_mol.conformers[0])  # type:ignore[index]
+    undef_mol.add_conformer(cis_mol.conformers[0])  # type:ignore
     assign_stereochemistry_from_3d(undef_mol)
     assert "Z" in [bond.stereochemistry for bond in undef_mol.bonds]
 
     undef_mol.generate_conformers(n_conformers=0)
 
-    undef_mol.add_conformer(trans_mol.conformers[0])  # type:ignore[index]
+    undef_mol.add_conformer(trans_mol.conformers[0])  # type:ignore
     assign_stereochemistry_from_3d(undef_mol)
     assert "E" in [bond.stereochemistry for bond in undef_mol.bonds]
 
@@ -127,12 +127,15 @@ def test_assign_stereochemistry_from_3d_atoms():
     r_mol.generate_conformers(n_conformers=1)
     s_mol.generate_conformers(n_conformers=1)
 
-    undef_mol.add_conformer(r_mol.conformers[0])  # type:ignore[index]
+    assert r_mol.conformers is not None
+    assert s_mol.conformers is not None
+
+    undef_mol.add_conformer(r_mol.conformers[0])
     assign_stereochemistry_from_3d(undef_mol)
     assert undef_mol.atom(0).stereochemistry == "R"
 
     undef_mol.generate_conformers(n_conformers=0)
 
-    undef_mol.add_conformer(s_mol.conformers[0])  # type:ignore[index]
+    undef_mol.add_conformer(s_mol.conformers[0])
     assign_stereochemistry_from_3d(undef_mol)
     assert undef_mol.atom(0).stereochemistry == "S"

--- a/openff/pablo/_utils.py
+++ b/openff/pablo/_utils.py
@@ -26,6 +26,7 @@ __all__ = [
     "cryst_to_box_vectors",
     "assign_stereochemistry_from_3d",
     "__UNSET__",
+    "dbg",
 ]
 
 T = TypeVar("T")
@@ -38,6 +39,13 @@ CIFValue: TypeAlias = str | float | int
 
 class __UNSET__:
     pass
+
+
+def dbg(o: T, msg: str = "{}") -> T:
+    if "{}" not in msg:
+        msg += ": {}"
+    print(msg.format(o))
+    return o
 
 
 def default_dict(

--- a/openff/pablo/ccd/__init__.py
+++ b/openff/pablo/ccd/__init__.py
@@ -11,11 +11,13 @@ from pathlib import Path
 from . import patches
 from ._ccdcache import CcdCache
 from .patches import (
+    add_dephosphorylated_5p_terminus,
     add_disulfide_crosslink,
     add_protonation_variants,
     add_synonyms,
     disambiguate_alt_ids,
     fix_caps,
+    set_hop3_leaving,
 )
 
 __all__ = [
@@ -35,6 +37,26 @@ CCD_RESIDUE_DEFINITION_CACHE: CcdCache = CcdCache(
             "CYS": add_disulfide_crosslink,
         },
         {"*": add_protonation_variants},
+        {
+            "U": add_dephosphorylated_5p_terminus,
+            "G": add_dephosphorylated_5p_terminus,
+            "C": add_dephosphorylated_5p_terminus,
+            "A": add_dephosphorylated_5p_terminus,
+            "DT": add_dephosphorylated_5p_terminus,
+            "DG": add_dephosphorylated_5p_terminus,
+            "DC": add_dephosphorylated_5p_terminus,
+            "DA": add_dephosphorylated_5p_terminus,
+        },
+        {
+            "U": set_hop3_leaving,
+            "G": set_hop3_leaving,
+            "C": set_hop3_leaving,
+            "A": set_hop3_leaving,
+            "DT": set_hop3_leaving,
+            "DG": set_hop3_leaving,
+            "DC": set_hop3_leaving,
+            "DA": set_hop3_leaving,
+        },
         {"*": disambiguate_alt_ids},
         {"*": add_synonyms},
     ],

--- a/openff/pablo/ccd/_ccdcache.py
+++ b/openff/pablo/ccd/_ccdcache.py
@@ -6,7 +6,7 @@ from urllib.request import urlopen
 
 from openmm.app.internal.pdbx.reader.PdbxReader import PdbxReader
 
-from ..chem import PEPTIDE_BOND
+from ..chem import PEPTIDE_BOND, PHOSPHODIESTER_BOND
 from ..residue import (
     AtomDefinition,
     BondDefinition,
@@ -263,9 +263,9 @@ LINKING_TYPES: dict[str, BondDefinition | None] = {
     # "D-saccharide, beta linking".upper(): [],
     # "DNA OH 3 prime terminus".upper(): [],
     # "DNA OH 5 prime terminus".upper(): [],
-    # "DNA linking".upper(): [],
-    # "L-DNA linking".upper(): [],
-    # "L-RNA linking".upper(): [],
+    "DNA linking".upper(): PHOSPHODIESTER_BOND,
+    "L-DNA linking".upper(): PHOSPHODIESTER_BOND,
+    "L-RNA linking".upper(): PHOSPHODIESTER_BOND,
     # "L-beta-peptide, C-gamma linking".upper(): [],
     # "L-gamma-peptide, C-delta linking".upper(): [],
     # "L-peptide COOH carboxy terminus".upper(): [],
@@ -276,7 +276,7 @@ LINKING_TYPES: dict[str, BondDefinition | None] = {
     # "L-saccharide, beta linking".upper(): [],
     # "RNA OH 3 prime terminus".upper(): [],
     # "RNA OH 5 prime terminus".upper(): [],
-    # "RNA linking".upper(): [],
+    "RNA linking".upper(): PHOSPHODIESTER_BOND,
     "non-polymer".upper(): None,
     # "other".upper(): [],
     "peptide linking".upper(): PEPTIDE_BOND,

--- a/openff/pablo/ccd/_ccdcache.py
+++ b/openff/pablo/ccd/_ccdcache.py
@@ -11,7 +11,7 @@ from ..residue import (
     AtomDefinition,
     BondDefinition,
     ResidueDefinition,
-    _defer_residue_definition_validation,
+    _skip_residue_definition_validation,
 )
 
 __all__ = [
@@ -65,10 +65,18 @@ class CcdCache(Mapping[str, list[ResidueDefinition]]):
         ] = [dict(d) for d in patches]
 
         for file in path.glob("*.cif"):
-            self._add_definition_from_str(file.read_text())
+            try:
+                self._add_definition_from_str(file.read_text())
+            except Exception:
+                # If adding a file fails, skip it - we want an error at runtime, not importtime
+                pass
 
         for resname in set(preload) - set(self._definitions):
-            self[resname]
+            try:
+                self[resname]
+            except Exception:
+                # If a preload fails, skip it - we want an error at runtime, not importtime
+                pass
 
     def __repr__(self):
         return (
@@ -95,7 +103,7 @@ class CcdCache(Mapping[str, list[ResidueDefinition]]):
         self,
         residue_definition: ResidueDefinition,
     ) -> list[ResidueDefinition]:
-        with _defer_residue_definition_validation():
+        with _skip_residue_definition_validation():
             definitions: list[ResidueDefinition] = [residue_definition]
             for patch_dict in self._patches:
                 patched_definitions: list[ResidueDefinition] = []
@@ -214,7 +222,7 @@ class CcdCache(Mapping[str, list[ResidueDefinition]]):
             else:
                 bonds = []
 
-            with _defer_residue_definition_validation():
+            with _skip_residue_definition_validation():
                 residue_definition = ResidueDefinition(
                     residue_name=residueName,
                     description=residue_description,

--- a/openff/pablo/ccd/patches.py
+++ b/openff/pablo/ccd/patches.py
@@ -2,8 +2,6 @@
 Patches to add essential features to the CCD.
 """
 
-import dataclasses
-from copy import deepcopy
 from itertools import combinations
 
 from openff.pablo.chem import DISULFIDE_BOND
@@ -25,6 +23,7 @@ __all__ = [
     "add_synonyms",
     "disambiguate_alt_ids",
     "add_disulfide_crosslink",
+    "add_dephosphorylated_5p_terminus",
 ]
 
 
@@ -50,6 +49,14 @@ ACIDIC_PROTONS: dict[str, list[str]] = {
     "TRP": ["HXT", "H2", "HE1"],
     "TYR": ["HXT", "H2", "HH"],
     "VAL": ["HXT", "H2"],
+    "DG": ["HOP2"],
+    "DA": ["HOP2"],
+    "DT": ["HOP2"],
+    "DC": ["HOP2"],
+    "G": ["HOP2"],
+    "A": ["HOP2"],
+    "U": ["HOP2"],
+    "C": ["HOP2"],
 }
 """Map from residue name to a list of atom names of abstractable hydrogens.
 
@@ -103,12 +110,10 @@ def fix_caps(res: ResidueDefinition) -> list[ResidueDefinition]:
     """
 
     return [
-        dataclasses.replace(
-            res,
+        res.replace(
             linking_bond=PEPTIDE_BOND,
             atoms=[
-                dataclasses.replace(
-                    atom,
+                atom.replace(
                     leaving=True if atom.name == "H" else atom.leaving,
                 )
                 for atom in res.atoms
@@ -145,7 +150,9 @@ def add_deprotonated_variants(res: ResidueDefinition) -> list[ResidueDefinition]
                 bonded_atoms.append(bond.atom1)
 
         if len(bonded_atoms) != 1:
-            raise ValueError("should be exactly 1 bonded atom to abstracted proton")
+            raise ValueError(
+                f"should be exactly 1 bonded atom to abstracted proton {hydrogen}, found {len(bonded_atoms)} in {res.description}",
+            )
         deprotonations.append((hydrogen, bonded_atoms[0]))
 
     variants: list[ResidueDefinition] = [res]
@@ -162,7 +169,7 @@ def add_deprotonated_variants(res: ResidueDefinition) -> list[ResidueDefinition]
             atoms: list[AtomDefinition] = []
             for atom in res.atoms:
                 if atom.name in partners:
-                    atoms.append(dataclasses.replace(atom, charge=atom.charge - 1))
+                    atoms.append(atom.replace(charge=atom.charge - 1))
                 elif atom.name in hydrogens:
                     if atom.symbol != "H":
                         raise ValueError(
@@ -172,8 +179,7 @@ def add_deprotonated_variants(res: ResidueDefinition) -> list[ResidueDefinition]
                     atoms.append(atom)
 
             variants.append(
-                dataclasses.replace(
-                    res,
+                res.replace(
                     atoms=atoms,
                     bonds=bonds,
                     description=res.description + f" -{' -'.join(hydrogens)}",
@@ -216,12 +222,11 @@ def add_protonated_variants(res: ResidueDefinition) -> list[ResidueDefinition]:
                 )
                 for i, atom in enumerate(res.atoms):
                     if atom.name == heavy_atom:
-                        atoms[i] = dataclasses.replace(atom, charge=atom.charge + 1)
+                        atoms[i] = atom.replace(charge=atom.charge + 1)
 
             hydrogens, _partners = zip(*combination)
             variants.append(
-                dataclasses.replace(
-                    res,
+                res.replace(
                     atoms=atoms,
                     bonds=bonds,
                     description=res.description + f" +{' +'.join(hydrogens)}",
@@ -235,26 +240,23 @@ def add_synonyms(res: ResidueDefinition) -> list[ResidueDefinition]:
     """
     Patch a residue definition to include synonyms from :py:data:`ATOM_NAME_SYNONYMS`.
     """
-    return [
-        dataclasses.replace(
-            res,
-            atoms=tuple(
-                dataclasses.replace(
-                    atom,
-                    synonyms=tuple(
-                        {
-                            *atom.synonyms,
-                            *ATOM_NAME_SYNONYMS.get(res.residue_name, {}).get(
-                                atom.name,
-                                [],
-                            ),
-                        },
-                    ),
+    patched_residues = [
+        res.replace(
+            atoms=(
+                atom.replace(
+                    synonyms={
+                        *atom.synonyms,
+                        *ATOM_NAME_SYNONYMS.get(res.residue_name, {}).get(
+                            atom.name,
+                            [],
+                        ),
+                    },
                 )
                 for atom in res.atoms
             ),
         ),
     ]
+    return patched_residues
 
 
 def disambiguate_alt_ids(res: ResidueDefinition) -> list[ResidueDefinition]:
@@ -283,7 +285,7 @@ def disambiguate_alt_ids(res: ResidueDefinition) -> list[ResidueDefinition]:
             if synonym in canonical_names:
                 clashes.append(i)
 
-    if clashes:
+    if len(clashes) != 0:
         old_to_new: dict[str, str] = {}
         for atom in res.atoms:
             if atom.synonyms:
@@ -291,48 +293,45 @@ def disambiguate_alt_ids(res: ResidueDefinition) -> list[ResidueDefinition]:
             else:
                 old_to_new[atom.name] = atom.name
 
-        res1 = dataclasses.replace(
-            res,
+        res1 = res.replace(
             atoms=[
-                dataclasses.replace(
-                    atom,
-                    synonyms=[] if i in clashes else deepcopy(atom.synonyms),
+                atom.replace(
+                    synonyms=[],
                 )
-                for i, atom in enumerate(res.atoms)
+                for atom in res.atoms
             ],
         )
-        res2 = dataclasses.replace(
-            res,
+        res2 = res.replace(
             atoms=[
-                dataclasses.replace(
-                    atom,
+                atom.replace(
                     name=old_to_new[atom.name],
-                    synonyms=[] if atom.synonyms else deepcopy(atom.synonyms),
+                    synonyms=[],
                 )
                 for atom in res.atoms
             ],
             bonds=[
-                dataclasses.replace(
-                    bond,
+                bond.replace(
                     atom1=old_to_new[bond.atom1],
                     atom2=old_to_new[bond.atom2],
                 )
                 for bond in res.bonds
             ],
-            description=res.description + "altids",
+            description=res.description + " altids",
             crosslink=(
                 None
                 if res.crosslink is None
-                else dataclasses.replace(
-                    res.crosslink,
+                else res.crosslink.replace(
                     atom1=old_to_new[res.crosslink.atom1],
+                    atom2=old_to_new.get(
+                        res.crosslink.atom2,
+                        res.crosslink.atom2,
+                    ),
                 )
             ),
             linking_bond=(
                 None
                 if res.linking_bond is None
-                else dataclasses.replace(
-                    res.linking_bond,
+                else res.linking_bond.replace(
                     atom1=old_to_new[res.linking_bond.atom1],
                     atom2=old_to_new.get(
                         res.linking_bond.atom2,
@@ -341,6 +340,7 @@ def disambiguate_alt_ids(res: ResidueDefinition) -> list[ResidueDefinition]:
                 )
             ),
         )
+        # print(f"{res2.bonds=} {old_to_new=}")
         return [res1, res2]
     else:
         return [res]
@@ -353,12 +353,68 @@ def add_disulfide_crosslink(res: ResidueDefinition) -> list[ResidueDefinition]:
         )
 
     return [
-        dataclasses.replace(
-            res,
+        res.replace(
             crosslink=DISULFIDE_BOND,
             atoms=[
-                atom if atom.name != "HG" else dataclasses.replace(atom, leaving=True)
+                atom if atom.name != "HG" else atom.replace(leaving=True)
                 for atom in res.atoms
             ],
+        ),
+    ]
+
+
+def add_dephosphorylated_5p_terminus(res: ResidueDefinition) -> list[ResidueDefinition]:
+    phosphate_names = {"P", "HOP2", "HOP3", "OP1", "OP2", "OP3"}
+
+    return [
+        res,
+        res.replace(
+            atoms=(
+                AtomDefinition(
+                    name="HO5'",
+                    synonyms=(),
+                    symbol="H",
+                    aromatic=False,
+                    charge=0,
+                    leaving=False,
+                    stereo=None,
+                ),
+                *(atom for atom in res.atoms if atom.name not in phosphate_names),
+            ),
+            bonds=(
+                BondDefinition(
+                    atom1="O5'",
+                    atom2="HO5'",
+                    aromatic=False,
+                    order=1,
+                    stereo=None,
+                ),
+                *(
+                    bond
+                    for bond in res.bonds
+                    if bond.atom1 not in phosphate_names
+                    and bond.atom2 not in phosphate_names
+                ),
+            ),
+            description=(
+                res.description[:-17] + "-5'-PROTONATED"
+                if res.description.endswith("-5'-MONOPHOSPHATE")
+                else res.description + " 5' dephosphorylated"
+            ),
+        ),
+    ]
+
+
+def set_hop3_leaving(res: ResidueDefinition) -> list[ResidueDefinition]:
+    """The OP3 and HOP3 atoms in nucleic acid residues are both absent when the
+    residue forms part of a polymer, but only OP3 is marked as leaving; HOP3
+    becomes a disconnected fragment"""
+    # TODO: Replace this with disconnected fragment detection in ResidueDefinition._leaving_fragment_of()
+    return [
+        res.replace(
+            atoms=(
+                atom.replace(leaving=True) if atom.name == "HOP3" else atom
+                for atom in res.atoms
+            ),
         ),
     ]

--- a/openff/pablo/chem.py
+++ b/openff/pablo/chem.py
@@ -7,6 +7,7 @@ from openff.pablo.residue import BondDefinition
 __all__ = [
     "DISULFIDE_BOND",
     "PEPTIDE_BOND",
+    "PHOSPHODIESTER_BOND",
 ]
 
 DISULFIDE_BOND = BondDefinition(
@@ -20,6 +21,14 @@ DISULFIDE_BOND = BondDefinition(
 PEPTIDE_BOND = BondDefinition(
     atom1="C",
     atom2="N",
+    order=1,
+    aromatic=False,
+    stereo=None,
+)
+
+PHOSPHODIESTER_BOND = BondDefinition(
+    atom1="O3'",
+    atom2="P",
     order=1,
     aromatic=False,
     stereo=None,

--- a/openff/pablo/exceptions.py
+++ b/openff/pablo/exceptions.py
@@ -14,6 +14,8 @@ from typing import TYPE_CHECKING
 
 from openff.toolkit import Molecule
 
+from openff.pablo._utils import flatten
+
 if TYPE_CHECKING:
     from openff.pablo._pdb_data import PdbData, ResidueMatch
     from openff.pablo.residue import ResidueDefinition
@@ -47,14 +49,31 @@ class NoMatchingResidueDefinitionError(ValueError):
             residue_definitions = list(residue_database[res_name])
             if len(residue_definitions) == 0:
                 msg.append("  No residues in database for residue name {res_name}")
-            found_names = [data.name[i] for i in res_atom_idcs]
+            found_names = {data.name[i] for i in res_atom_idcs}
             for resdef in residue_definitions:
-                resdef_names = [
-                    "|".join([atom.name, *atom.synonyms]) for atom in resdef.atoms
-                ]
+                extra_names = found_names.difference(
+                    flatten([atom.name, *atom.synonyms] for atom in resdef.atoms),
+                )
+                missing_names = {
+                    "|".join([atom.name, *atom.synonyms])
+                    for atom in resdef.atoms
+                    if atom.name not in found_names
+                    and all([synonym not in found_names for synonym in atom.synonyms])
+                    and not atom.leaving
+                }
+                missing_leavers = {
+                    "|".join([atom.name, *atom.synonyms])
+                    for atom in resdef.atoms
+                    if atom.name not in found_names
+                    and all([synonym not in found_names for synonym in atom.synonyms])
+                    and atom.leaving
+                }
                 msg.append(f"    In {resdef.description}:")
-                msg.append(f"      Expected {sorted(resdef_names)}")
-                msg.append(f"      Found {sorted(found_names)}")
+                msg.append(f"      {sorted(extra_names)} were found but not expected")
+                msg.append(f"      {sorted(missing_names)} were expected but not found")
+                msg.append(
+                    f"      Leaving atoms {sorted(missing_leavers)} were also not found",
+                )
 
         super().__init__("\n".join(msg))
 

--- a/openff/pablo/residue.py
+++ b/openff/pablo/residue.py
@@ -13,7 +13,7 @@ from typing import Literal, Self
 from openff.toolkit import Molecule
 from openff.units import elements, unit
 
-from openff.pablo._utils import unwrap
+from openff.pablo._utils import __UNSET__, unwrap
 
 __all__ = [
     "AtomDefinition",
@@ -25,7 +25,7 @@ _residue_definition_skip_validation = False
 
 
 @contextmanager
-def _defer_residue_definition_validation():  # type: ignore[deadcode]
+def _skip_residue_definition_validation():  # type: ignore[deadcode]
     global _residue_definition_skip_validation
     _residue_definition_skip_validation = True
     yield
@@ -82,6 +82,28 @@ class AtomDefinition:
             stereo=stereo,
         )
 
+    def replace(
+        self,
+        name: str | __UNSET__ = __UNSET__(),
+        symbol: str | __UNSET__ = __UNSET__(),
+        synonyms: Iterable[str] | __UNSET__ = __UNSET__(),
+        leaving: bool | __UNSET__ = __UNSET__(),
+        charge: int | __UNSET__ = __UNSET__(),
+        aromatic: bool | __UNSET__ = __UNSET__(),
+        stereo: Literal["S", "R"] | None | __UNSET__ = __UNSET__(),
+    ) -> Self:
+        return self.__class__(
+            name=self.name if isinstance(name, __UNSET__) else name,
+            symbol=self.symbol if isinstance(symbol, __UNSET__) else symbol,
+            synonyms=(
+                self.synonyms if isinstance(synonyms, __UNSET__) else tuple(synonyms)
+            ),
+            leaving=self.leaving if isinstance(leaving, __UNSET__) else leaving,
+            charge=self.charge if isinstance(charge, __UNSET__) else charge,
+            aromatic=self.aromatic if isinstance(aromatic, __UNSET__) else aromatic,
+            stereo=self.stereo if isinstance(stereo, __UNSET__) else stereo,
+        )
+
 
 @dataclass(frozen=True)
 class BondDefinition:
@@ -119,6 +141,22 @@ class BondDefinition:
             order=order,
             aromatic=aromatic,
             stereo=stereo,
+        )
+
+    def replace(
+        self,
+        atom1: str | __UNSET__ = __UNSET__(),
+        atom2: str | __UNSET__ = __UNSET__(),
+        order: int | __UNSET__ = __UNSET__(),
+        aromatic: bool | __UNSET__ = __UNSET__(),
+        stereo: Literal["E", "Z"] | None | __UNSET__ = __UNSET__(),
+    ) -> Self:
+        return self.__class__(
+            atom1=self.atom1 if isinstance(atom1, __UNSET__) else atom1,
+            atom2=self.atom2 if isinstance(atom2, __UNSET__) else atom2,
+            order=self.order if isinstance(order, __UNSET__) else order,
+            aromatic=(self.aromatic if isinstance(aromatic, __UNSET__) else aromatic),
+            stereo=self.stereo if isinstance(stereo, __UNSET__) else stereo,
         )
 
 
@@ -199,6 +237,34 @@ class ResidueDefinition:
     bonds: tuple[BondDefinition, ...]
     """The bond definitions that make up this residue"""
 
+    def replace(
+        self,
+        residue_name: str | __UNSET__ = __UNSET__(),
+        description: str | __UNSET__ = __UNSET__(),
+        linking_bond: BondDefinition | None | __UNSET__ = __UNSET__(),
+        crosslink: BondDefinition | None | __UNSET__ = __UNSET__(),
+        atoms: Iterable[AtomDefinition] | __UNSET__ = __UNSET__(),
+        bonds: Iterable[BondDefinition] | __UNSET__ = __UNSET__(),
+    ) -> Self:
+        return self.__class__(
+            residue_name=(
+                self.residue_name
+                if isinstance(residue_name, __UNSET__)
+                else residue_name
+            ),
+            description=(
+                self.description if isinstance(description, __UNSET__) else description
+            ),
+            linking_bond=(
+                self.linking_bond
+                if isinstance(linking_bond, __UNSET__)
+                else linking_bond
+            ),
+            crosslink=self.crosslink if isinstance(crosslink, __UNSET__) else crosslink,
+            atoms=self.atoms if isinstance(atoms, __UNSET__) else tuple(atoms),
+            bonds=self.bonds if isinstance(bonds, __UNSET__) else tuple(bonds),
+        )
+
     def __post_init__(self):
         if _residue_definition_skip_validation:
             return
@@ -206,31 +272,37 @@ class ResidueDefinition:
         self._validate()
 
     def _validate(self):
-        if (
-            self.linking_bond is None
-            and self.crosslink is None
-            and True in {atom.leaving for atom in self.atoms}
-        ):
-            raise ValueError(
-                f"{self.residue_name}: Leaving atoms were specified, but there is no linking bond or crosslink",
-                self,
-            )
-        if len({atom.name for atom in self.atoms}) != len(self.atoms):
-            raise ValueError(
-                f"{self.residue_name}: All atoms must have unique canonical names",
-            )
+        try:
+            if (
+                self.linking_bond is None
+                and self.crosslink is None
+                and True in {atom.leaving for atom in self.atoms}
+            ):
+                raise ValueError(
+                    f"{self.residue_name}: Leaving atoms were specified, but there is no linking bond or crosslink",
+                    self,
+                )
+            if len({atom.name for atom in self.atoms}) != len(self.atoms):
+                raise ValueError(
+                    f"{self.residue_name}: All atoms must have unique canonical names",
+                )
 
-        all_leaving_atoms = {atom.name for atom in self.atoms if atom.leaving}
-        assigned_leaving_atoms = (
-            self.prior_bond_leaving_atoms
-            | self.posterior_bond_leaving_atoms
-            | self.crosslink_leaving_atoms
-        )
-        unassigned_leaving_atoms = all_leaving_atoms.difference(assigned_leaving_atoms)
-        if len(unassigned_leaving_atoms) != 0:
-            raise ValueError(
-                f"{self.residue_name}: Leaving atoms could not be assigned to a bond: {unassigned_leaving_atoms}",
+            all_leaving_atoms = {atom.name for atom in self.atoms if atom.leaving}
+            assigned_leaving_atoms = (
+                self.prior_bond_leaving_atoms
+                | self.posterior_bond_leaving_atoms
+                | self.crosslink_leaving_atoms
             )
+            unassigned_leaving_atoms = all_leaving_atoms.difference(
+                assigned_leaving_atoms,
+            )
+            if len(unassigned_leaving_atoms) != 0:
+                raise ValueError(
+                    f"{self.residue_name}: Leaving atoms could not be assigned to a bond: {unassigned_leaving_atoms}",
+                )
+        except KeyError as e:
+            # print(self)
+            raise e
 
     @classmethod
     def from_molecule(
@@ -512,19 +584,27 @@ class ResidueDefinition:
                 yield bond.atom1
 
     def _leaving_fragment_of(self, linking_atom: str) -> Iterator[str]:
+        # print(f"{self=} {linking_atom=}")
         atoms_to_check = list(self.atoms_bonded_to(linking_atom))
         checked_atoms: set[str] = set()
         while atoms_to_check:
             atom_name = atoms_to_check.pop()
+            # print(f"{atoms_to_check=} {checked_atoms=} {atom_name=}")
             if self.name_to_atom[atom_name].leaving:
+                # print(1)
                 yield atom_name
+                # print(2)
                 atoms_to_check.extend(
                     filter(
                         lambda x: x not in checked_atoms,
                         self.atoms_bonded_to(atom_name),
                     ),
                 )
+                # print(3)
             checked_atoms.add(atom_name)
+            # print(4)
+
+        # print(5)
 
     @cached_property
     def posterior_bond_leaving_atoms(self) -> set[str]:

